### PR TITLE
added event subscriptions field

### PR DESCRIPTION
**Pull request by davidhammturner**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
Add passing integration test logs here
```
</details>

# Example query results
<details>
  <summary>Results</summary>
> select name,event_subscriptions from aws_inspector_assessment_template where name like '%POC%' or name like '%caleb%';;

                                                                                                     

```
Add example SQL query results here (please include the input queries as well)
```
+----------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------
| name                       | event_subscriptions                                                                                                                                        
+----------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------
| Saad-POC-Template          | [{"EventSubscriptions":[{"Event":"FINDING_REPORTED","SubscribedAt":"2020-07-24T12:25:48.06Z"}],"ResourceArn":"arn:aws:inspector:us-east-1:055850966408:targ
| wmcso-inspector-caleb-test | []                                                                                                                                                         
+----------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------
(END)
</details>

### CODE REVIEW

File: aws/table_aws_inspector_assessment_template.go [hunk 1]
```diff
@@ -220,6 +226,40 @@ func getAwsInspectorAssessmentTemplateTags(ctx context.Context, d *plugin.QueryD
...
+		return nil, err
+	}
+
+	// Build the params
+	params := &inspector.ListEventSubscriptionsInput{
+		ResourceArn: &assessmentTemplateArn,
+	}
+
+	// Get call
+	op, err := svc.ListEventSubscriptions(params)
```
Comment (cbruno10) [inline aws/table_aws_inspector_assessment_template.go:254]: @davidhammturner Can you please try using the `ListEventSubscriptionsPages` method to ensure that we get all of the subscriptions? There's an example in https://github.com/turbot/steampipe-plugin-aws/blob/main/aws/table_aws_iam_role.go#L254.

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@davidhammturner Please see my review comment and let me know if you have any questions!

**Comment by davidhammturner** [comment]
Sure, I'll do that. Sorry for the troubles.



David Hamm
Sr. Cloud Security Engineer
O 404-827-2947          C 404-451-4824

[signature_1065577666]

From: cbruno10 ***@***.***>
Date: Tuesday, June 1, 2021 at 4:53 PM
To: turbot/steampipe-plugin-aws ***@***.***>
Cc: Hamm, David ***@***.***>, Mention ***@***.***>
Subject: Re: [turbot/steampipe-plugin-aws] added event subscriptions field (#467)

@cbruno10 requested changes on this pull request.

@davidhammturner<https://urldefense.com/v3/__https:/github.com/davidhammturner__;!!AQdq3sQhfUj4q8uUguY!x6CudnaHXEZm9n75y9H-kBPSVVMV7e1RCRGuFWEcGTzcug_62lvoSS4sXG2hsWDl$> Please see my review comment and let me know if you have any questions!

________________________________

In aws/table_aws_inspector_assessment_template.go<https://urldefense.com/v3/__https:/github.com/turbot/steampipe-plugin-aws/pull/467*discussion_r643471318__;Iw!!AQdq3sQhfUj4q8uUguY!x6CudnaHXEZm9n75y9H-kBPSVVMV7e1RCRGuFWEcGTzcug_62lvoSS4sXMnFB6ep$>:

> +

+ assessmentTemplateArn := *h.Item.(*inspector.AssessmentTemplate).Arn

+

+ // Create Session

+ svc, err := InspectorService(ctx, d, region)

+ if err != nil {

+        return nil, err

+ }

+

+ // Build the params

+ params := &inspector.ListEventSubscriptionsInput{

+        ResourceArn: &assessmentTemplateArn,

+ }

+

+ // Get call

+ op, err := svc.ListEventSubscriptions(params)

@davidhammturner<https://urldefense.com/v3/__https:/github.com/davidhammturner__;!!AQdq3sQhfUj4q8uUguY!x6CudnaHXEZm9n75y9H-kBPSVVMV7e1RCRGuFWEcGTzcug_62lvoSS4sXG2hsWDl$> Can you please try using the ListEventSubscriptionsPages method to ensure that we get all of the subscriptions? There's an example in https://github.com/turbot/steampipe-plugin-aws/blob/main/aws/table_aws_iam_role.go#L254<https://urldefense.com/v3/__https:/github.com/turbot/steampipe-plugin-aws/blob/main/aws/table_aws_iam_role.go*L254__;Iw!!AQdq3sQhfUj4q8uUguY!x6CudnaHXEZm9n75y9H-kBPSVVMV7e1RCRGuFWEcGTzcug_62lvoSS4sXFtnoWGZ$>.

**Comment by cbruno10** [comment]
@davidhammturner No trouble at all, we greatly appreciate the contribution!

**Comment by davidhammturner** [comment]
Hi,

I switched the call to ListEventSubscriptionsPages.  However this brought up a couple questions.
1.  Since the plugin converts the field name event_subscriptions to use as transform key, named EventSubscriptions, I had to allocate a struct to return the data with the required key.  Did I do this correctly?
2. The data returned is an array of objects, each containing an EventSubscriptions object.  This suggests the data should be represented in an additional table like aws_inspector_template_eventsubscriptons.  However, the API call requires a resource ARN to list the subscriptons.  How do you populate a table that relies on a field from another table? 

Thanks.

**Comment by cbruno10** [comment]
@johnsmyth For template assessment event subscriptions, it seems like the format from AWS isn't a perfect fit for our column out of the box:
```
[{
	"EventSubscriptions": [{
		"Event": "ASSESSMENT_RUN_COMPLETED",
		"SubscribedAt": "2021-06-09T13:23:40.073Z"
	}, {
		"Event": "ASSESSMENT_RUN_STARTED",
		"SubscribedAt": "2021-06-09T13:23:40.072Z"
	}, {
		"Event": "ASSESSMENT_RUN_STATE_CHANGED",
		"SubscribedAt": "2021-06-09T13:23:40.072Z"
	}, {
		"Event": "FINDING_REPORTED",
		"SubscribedAt": "2020-08-10T19:18:00.997Z"
	}],
	"ResourceArn": "arn:aws:inspector:us-east-1:123456789408:target/0-LPFe3n2v/template/0-qGrEluCz",
	"TopicArn": "arn:aws:sns:us-east-1:123456789408:test-inspector-us-east-1"
}, {
	"EventSubscriptions": [{
		"Event": "FINDING_REPORTED",
		"SubscribedAt": "2020-08-10T19:18:00.997Z"
	}],
	"ResourceArn": "arn:aws:inspector:us-east-1:123456789408:target/0-LPFe3n2v/template/0-qGrEluCz",
	"TopicArn": "arn:aws:sns:us-east-1:123456789408:test-inspector-collector-us-east-1"
}]
```

Using the column name `event_subscriptions` seemed like the best fit at first based on the API call, but now with that column name the `EventSubscriptions` field in the data seems redundant. Would you still call the column `event_subscriptions`? 

Also, should we keep the `ResourceArn` attribute, which will always be the same in each item since we're hydrating this data per template assessment/row?

Do you know of any other examples where we've received data like this before?

**Comment by johnsmyth** [comment]
@cbruno10 while it does appear redundant, `event_subscriptions` is probably the best name for the column - thats what a user would likely be looking for.  As far as the data in the column,  we generally do NOT modify the source format to our own arbitrary format when the column is json, but instead use the API json.  Doing so would open up a whole mess of arbitrary decisions, and at the end we would end up with an undocumented schema, as opposed to the cloud providers native schema which is typically documented and understood.  So I would keep the native format, including the `ResourceArn` attribute.  It is also an option to add the data into a separate table with a List KeyColumn, as @davidhammturner suggested, and there are certainly trade offs there in performance and flexibity that depend on the typical usage and number of rows.  Im not familiar enough with this particular api / resource to make a recommendation but happy to discuss with you.

**Comment by davidhammturner** [comment]
I've spent some time working with queries and the event_subscriptions column.  Since Inspector event subscriptions cannot be listed without supplying a resourceARN we should probably leave this as a column in the aws_inspector_template table.  Breaking it out into an additional table seems to add little value.  Requiring the use of jsonb_array_elements to extract the data appears as if it will be common among many services and offer some consistency in querying many services.

File: aws/table_aws_inspector_assessment_template.go [hunk 2]
```diff
@@ -86,7 +86,12 @@ func tableAwsInspectorAssessmentTemplate(_ context.Context) *plugin.Table {
 				Hydrate:     getAwsInspectorAssessmentTemplateTags,
 				Transform:   transform.FromField("Tags"),
 			},
-
+			{
+				Name:        "event_subscriptions",
+				Description: "A list of event subscriptions associated with the Assessment Template.",
+				Type:        pb.ColumnType_JSON,
+				Hydrate:     listAwsInspectorAssessmentEventSubscriptions,
```
Comment (cbruno10) [inline aws/table_aws_inspector_assessment_template.go:93]:
```suggestion
				Hydrate:     listAwsInspectorAssessmentEventSubscriptions,
                              Transform:   transform.FromValue()
```

File: aws/table_aws_inspector_assessment_template.go [hunk 3]
```diff
@@ -220,6 +225,50 @@ func getAwsInspectorAssessmentTemplateTags(ctx context.Context, d *plugin.QueryD
...
+		return nil, err
+	}
+
+	// Build the params
+	params := &inspector.ListEventSubscriptionsInput{
+		ResourceArn: &assessmentTemplateArn,
+	}
+	// Create a Struct name EventSubscriptions as the framework will use the field name to create
+	// a key from which it extracts the contents of the field
+	type AssociatedEventSubscriptions struct {
```
Comment (cbruno10) [inline aws/table_aws_inspector_assessment_template.go:253]: @davidhammturner For your earlier question, we don't need a `struct` here; instead you can store the `[]*inspector.Subscription` data in a variable and return that, e.g.,
```
var eventSubscriptions []*inspector.Subscription
```

I've also included a suggestion in the column definition that uses `transform.FromValue`, which tells the column to just use the returned data (more information on these transform functions can be found at https://steampipe-io-git-staging-turbot.vercel.app/docs/develop/writing-plugins#transform-functions).

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@davidhammturner I've included some suggested changes, please have a look and let me know if you have any questions. Thanks!

**Comment by cbruno10** [comment]
> I've spent some time working with queries and the event_subscriptions column. Since Inspector event subscriptions cannot be listed without supplying a resourceARN we should probably leave this as a column in the aws_inspector_template table. Breaking it out into an additional table seems to add little value. Requiring the use of jsonb_array_elements to extract the data appears as if it will be common among many services and offer some consistency in querying many services.

@davidhammturner Thanks for the info - I agree with your assessment that this information should remain as a column in the `aws_inspector_template` table.

**Review by cbruno10 — APPROVED** [review]

**Comment by cbruno10** [comment]
@davidhammturner PR looks great, thanks again for the contribution!

**Pull request merged**